core/src/main/scala/org/http4s/multipart/MultipartDecoder.scala

```diff
@@ -0,0 +1,62 @@
+package org.http4s
+package multipart
+
+import scala.util.{Try, Success, Failure}
+import scala.util.control._
+import org.http4s._
+import parser._
+import headers._
+import Http4s._
+import org.http4s.util._
+import scalaz.concurrent._
+import scalaz.concurrent.Task._
+
+import scodec.bits.ByteVector
+import scalaz._
+import Scalaz._
+import scalaz.stream._
+import scalaz.stream.Process._
+import org.parboiled2._
+import org.log4s.getLogger
+
+private[http4s] object MultipartDecoder {
+
+  private[this] val logger = getLogger
+  
+  val decoder: EntityDecoder[Multipart] =
+    EntityDecoder.decodeBy(MediaRange.`multipart/*`) { msg =>
+      def gatherParts = {
+        def go(part: Part): Process1[Headers \/ ByteVector, Part] =
+          receive1Or[Headers \/ ByteVector, Part](emit(part)) {
+            case -\/(headers) =>
+              emit(part) fby go(Part(headers, EmptyBody))
+            case \/-(chunk) =>
+              go(part.copy(body = part.body ++ emit(chunk)))
```

Comment:
This is where we want to split off and create a temp-file backed Part, if the part has grown too large.  (commons-fileupload uses a 10kB threshold by default).

This is not as easy as it sounds: writing to a temp file is effectful.  This suggests a channel, but channels emit one Task per input.  This is probably going to result in Option spup.  The other alternative I see is to brazenly block on java.io side effects and be very careful about finally blocks.